### ISSUE TITLE
Linux: Do not use header bar by default on Wayland

### ISSUE THREAD

**Issue by hep-mh** [flutter/flutter#111453]
By default, Linux applications use a GTK 3 header bar instead of a traditional title bar when running on GNOME, but also more generally when running on Wayland (independently of the desktop environment). Hence, a KDE Plasma session on Wayland uses a header bar, which looks distinctively different from the window decorations of the rest of the system. To illustrate this point, below I show the titelbar for the system files app, an example Flutter app on X11, and an example Flutter app on Wayland, on my KDE neon machine, running KDE Plasma 5.25.

![neon_files](https://user-images.githubusercontent.com/47532112/189707700-c0850775-4957-4e0d-84e5-3eaf67c04084.png)
![neon_x11](https://user-images.githubusercontent.com/47532112/189707727-06109a24-39ff-460a-a3ce-bcca75ce7fda.png)
![neon_wayland](https://user-images.githubusercontent.com/47532112/189707748-72c611e4-fd18-4bea-8ba3-ea39f349734d.png)

Evidently, the Flutter titlebar on X11 (second image) matches the general style of the system files app (first image). However, the same is not true for the Flutter app on Wayland (third image). In this case, the title is centered, the app icon is missing, the height of the titelbar is too large, and additional icons next to the close button etc. (which are chosen by the user) are missing. **This problem can be mitigated by using a traditional titelbar instead of a header bar in this Wayland environment, which I have tested by modifying the code in ``linux/my_application.cc``.** I therefore advocate for not using a header bar on Wayland by default, but only when running in a Gnome environment.

As mentioned above, I am aware of the fact that I can manually disable this behavior in ``linux/my_application.cc``, and this is exactly what I am doing for my app. However, every Flutter app, which uses the default behavior will lead to a subpar experience on Plasma Wayland, which should not be the default behavior.

Moreover, there are also problems when using a header bar on GNOME (Wayland and X11), especially when packaging the app as a flatpak. I can elaborate more on this if requested. However, I have tested that a traditional titelbar looks good in every environment. Hence, it might even be an option to disable the header bar by default on every system and only enable it via plugins if it is really needed, i.e. if elements like buttons are meant to be placed in the header bar. Otherwise it does not have a purpose anyway, as far as I am aware.

**Comment by exaby73** [comment]
Hello @hep-mh. Thank you for filing this issue. Labelling for further insights from the team.

**Comment by hep-mh** [comment]
I agree. Implementing #94381 and #31373 should also solves this issue.

**Comment by ArikRahman** [comment]
Howdy,

I would support not having the title bar on by default, especially for wayland environments. It is adding a title bar that clashes with the program.

Uploading 2024-06-08 19-25-35.mp4…

**Comment by ArikRahman** [comment]
I support this proposal since it is prescribed that in Gnome's open standard that it shouldn't be rendered by default, and would look especially out of place on wayland systems. Should be an argument like --no-title-bar to prevent flutter handling title bar.

https://github.com/flutter/flutter/assets/40479733/a08c525b-d027-478f-b428-48be935b9e23

**Comment by TechnicJelle** [comment]
I fixed this by editing the `linux/runner/my_application.cc` file.

`gboolean use_header_bar` is now `FALSE` by default, and only on GNOME, does it get set to `TRUE`.
The X11 check has barely changed, other than being inverted now, and I added a whole new check for on Wayland.
Apparently that check isn't as nice&simple on Wayland as on X11, so I had to add an extra convenience function.

You can see the full diff here: https://github.com/TechnicJelle/BlueMapGUI/commit/f566411761f56de8e612ed6bdfbafb0afcbd537e

Or directly as text here:
```diff
diff --git a/linux/runner/my_application.cc b/linux/runner/my_application.cc
index 0b7a4bf..03887c0 100644
--- a/linux/runner/my_application.cc
+++ b/linux/runner/my_application.cc
@@ -19,29 +19,50 @@ static void first_frame_cb(MyApplication* self, FlView* view) {
   gtk_widget_show(gtk_widget_get_toplevel(GTK_WIDGET(view)));
 }
 
+// Simple string contains check
+static gboolean str_contains_case_insensitive_ascii(const gchar* haystack, const gchar* needle) {
+  if (haystack == nullptr || needle == nullptr) return FALSE;
+
+  gchar* lower_haystack = g_ascii_strdown(haystack, -1);
+  gchar* lower_needle = g_ascii_strdown(needle, -1);
+  gboolean result = g_strrstr(lower_haystack, lower_needle) != nullptr;
+  g_free(lower_haystack);
+  g_free(lower_needle);
+  return result;
+}
+
 // Implements GApplication::activate.
 static void my_application_activate(GApplication* application) {
   MyApplication* self = MY_APPLICATION(application);
   GtkWindow* window =
       GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));
 
-  // Use a header bar when running in GNOME as this is the common style used
-  // by applications and is the setup most users will be using (e.g. Ubuntu
-  // desktop).
-  // If running on X and not using GNOME then just use a traditional title bar
-  // in case the window manager does more exotic layout, e.g. tiling.
-  // If running on Wayland assume the header bar will work (may need changing
-  // if future cases occur).
-  gboolean use_header_bar = TRUE;
+  // On normal WMs, use a traditional title bar. On GNOME, use a header bar as this is the common style on that WM.
+  gboolean use_header_bar = FALSE;
+
+  // Check for GNOME on X11
 #ifdef GDK_WINDOWING_X11
   GdkScreen* screen = gtk_window_get_screen(window);
   if (GDK_IS_X11_SCREEN(screen)) {
     const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
-    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
-      use_header_bar = FALSE;
+    if (g_strcmp0(wm_name, "GNOME Shell") == 0) {
+      use_header_bar = TRUE;
     }
   }
 #endif
+  // Check for GNOME on Wayland
+  if (use_header_bar == FALSE) {
+    const gchar* current_desktop = g_getenv("XDG_CURRENT_DESKTOP");
+    const gchar* session_desktop = g_getenv("XDG_SESSION_DESKTOP");
+    const gchar* desktop_session = g_getenv("DESKTOP_SESSION");
+    gboolean isGnome = str_contains_case_insensitive_ascii(current_desktop, "gnome") ||
+                       str_contains_case_insensitive_ascii(session_desktop, "gnome") ||
+                       str_contains_case_insensitive_ascii(desktop_session, "gnome");
+    if (isGnome) {
+      use_header_bar = TRUE;
+    }
+  }
+
   if (use_header_bar) {
     GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
     gtk_widget_show(GTK_WIDGET(header_bar));
```

Hope this helps someone! :)

**Comment by robert-ancell** [comment]
The decision to use a header bar was taken early in the Flutter Linux development and I think time has shown this is not been a good choice. The assumption was a header bar would work in all situations, but this has turned out to either be not the case (e.g. some tiling window managers) or undesirable (e.g. running in KDE). Instead of building ever more complex heuristics to decide which method to use I think the correct solution is the minimal case of not using a header bar and leaving it to the application developer to decide if they want to opt-in to one (as is the case for all GTK apps). Proposed in https://github.com/flutter/flutter/pull/190120
